README
This code computes sunrise and sunset times. There is a library and an
assortment of utilities. The most useful utility is x10events. This
program parses crontab files and updates the time fields to track sun
events given instructions in the comments.

Combined with an X10 interface like HEYU or X10 this allows the
scheduling lighting events that follow the changing seasons and DST
automatically.  This application does not depend on X10, it can be
used for any crontab entry you like.

It is a port of some BASIC code by Roger W. Sinnott for the August
1994 Sky & Telescope that I found on their website:
  http://www.skypub.com/resources/software/basic/programs/sunup.bas

I apologize for the inscrutability of the code, but I used the same opaque
variable names used in the BASIC code to make debugging during the port
more reasonable.

riseset.mat is a massaged version of riseset.txt: the official sunrise and
sunset times for Cambridge, Massachusetts, USA, 71:07W, 42:22N for the
year 1999 from the U.S. Naval Observatory website at:
  http://aa.usno.navy.mil/AA/data/docs/RS_OneYear.html

There are several programs: today, year, x10events.

'today' simply prints the time and azimuth of today's sunrise and sunset. 

'year' prints an entire year's worth of sunrise and sunset times.  The
companion script year.sh uses year, the table riseset.mat, and gnuplot to
generate a verification plot.

'x10events' is really the whole point of all this work.  It takes a
crontab file that is assumed to contain trailing information about
when the events are supposed to happen relative to sunrise and sunset
events and generates a new table containing the correct times.  I use
this to update my heyu timers by placing the script x10events.sh in
/etc/cron.daily.  'x10events' assumes that all entries in the crontab
are events that it should manage - I suggest placing these events in a
separate crontab file - see the FAQ for a more detailed example.

All programs take three arguments: latitude longitude and timezone.
All values should be expressed as decimal values, not H:M:S.  Latitute
is positive North of the equator.  Longitude is positive East of the
prime meridian.  Timezones are positive East of UTC and should be
expressed in hours for Standard Time.  Do not include Daylight Savings
Time offsets.

If the arguments are omited the defaults "42.3778343 -71.1063232 -5" are
used.  Those correspond to Cambridge MA USA, EST.

The package has been tested and successfully builds and runs under
Redhat 5.2, 6.0, 6.2, and 7.2, OSF1 V4.0 (CXX=g++), IRIX64 6.4
(CXX=CC), and IRIX 6.5 (CXX=CC).

Christopher R. Wren
c@drwren.com



CHANGELOG
ToDo:
	* cronevents: support dawn and dusk like rise and set
	* cronevents: simple way to build with different tag (not "#x10")


0.52sdb
	* rename x10events => cronevents

0.51sdb
	* x10events: handle lines up to 2.5KBytes (was 256 bytes)

0.5sdb
	* bluack on the C++ style headers
	* merged with my other changes, cleaned up dir, etc.

0.5
	* x10events: include Doug's deprecated headers fix
	* rpm: added a spec file for rpms, tested on FC2

0.46sdb
	* x10events: tweak summary crontab comments
	* year: -v option to output day of year, -v -v for debug output

0.45sdb
	* x10events: tweak summary crontab comments
	* suntime.c: added
	* ephem.cc: corrected a comment typo

0.44sdb
	* x10events: add/revise header documentation
	* x10events: tweak summary crontab comments

0.43sdb
	* x10events: fix: again support decimal hours (eg 1.1=1:06)
	* x10events: preserve crontab's original "mean" and "variance" strings

0.42sdb
	* x10events: new: riseif and setif conditionals
	* x10events: fix: calc times as integer minutes (not floating point)

0.41sdb
	* all: timezone from environment, not specified
	* all: default long-lat from Makefile
	* all: version macro in Makefile
	* x10events: more robust crontab sharing
	* x10events: append summary comments to crontab

0.4sdb
	* all: compile cleanly with gcc 3.3

0.4
	* x10events: fixed dropped zero mean in the presence of non-zero var
	* x10events: cleaned up -pedantic warnings

0.3  
	* x10events: randomized events 
	* x10events: preserve day month and dow fields
	* x10events: preserve whole-line comments 	             
	* x10events: look in environment for LAT LON and TOFF 
	* x10events: read/write to a file to interoperate with crontab -e
	* x10events: accept [-]hh:mm format in addition to decimal hours
	* x10events.sh: use "crontab -e" instead of hokey "crontab -l" hack

0.2
	* x10events: fixed crontab "comment creep" caused by x10events 
	* x10events: pad hours and minutes with leading 0's 
	* added a FAQ. 

0.1
	* Initial 

thanks to

	"Doug Lim" <doug.lim@gmail.com>
	"Phil Gregory" <phil_g@pobox.com>
	"Russell Salerno" <rsalerno@li.net>
	"Matt Williams" <mattrwilliams@yahoo.com>


cronevents.cc
// cronevents.cc: adjust x10 crontab according to comments
// Copyright (C) 1999 "Christpher R. Wren" <c@drwren.com>
// 
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

// by "Christpher R. Wren" <c@drwren.com>
// Sun Oct 10 1999 

// interpret crontab comments and modify the entry accordingly
// comments should be #x10 {token} [offset [variance]]
// where {token} is one of:
//   abs		for absolute times, see note below
//   rise		for times relative to sunrise
//   set		for times relative to sunset
//   riseif before	conditional if time is before sunrise, see note
//   riseif after	conditional if time is after sunrise, see note
//   setif before	conditional if time is before sunset, see note
//   setif after	conditional if time is after sunset, see note
// and:
//  offset		specifies time relative to the token
//  variance		specifies a bound for uniform noise added to the time
//  offset / variance	may be specified as decimal hours or as [-]hh:mm
//
// note for abs and riseif/setif:
//  * If offset is specified, then the offset value is the mean/base time.
//  * If no arguments, then the crontab hour and minute fields are the time.
//
// for example:
// 
// execute at sunrise
//   11 07 * * * /usr/local/bin/heyu turn tree on        #x10 rise
// execute at 6 minutes (0.1 hour) after sunset
//   36 16 * * * /usr/local/bin/heyu turn tree off       #x10 set 0.1
//   36 16 * * * /usr/local/bin/heyu turn tree off       #x10 set 0:06
// execute at 30 minutes (0.5 hour) before sunset
//   04 16 * * * /usr/local/bin/heyu turn floorlamp on   #x10 set -0.50
// execute at midnight
//   00 00 * * * /usr/local/bin/heyu turn floorlamp off  #x10 abs
// execute at 30 minutes after midnight, +/- 30 minutes
//   12 00 * * * /usr/local/bin/heyu turn office off     #x10 abs 00:30 0.5
// execute at 7am, if that is before sunrise
//   00 07 * * * /usr/local/bin/heyu turn overhead on    #x10 riseif before
//   99 99 * * * /usr/local/bin/heyu turn overhead on    #x10 riseif before 7:00
// execute at 7am +/- 15 minutes, if that is before sunrise
//   99 99 * * * /usr/local/bin/heyu turn overhead on    #x10 riseif before 7:00 0:15


/**********************************************************
 * ToDo:
 *   resolve issue with crontab(5):
 *     "comments are not allowed on the same line as a command"
 */

#include <iostream>
#include <sstream>
#include <fstream>
#include <cmath>
#include <cstdlib>
#include <unistd.h>
#include <ctime>
#include <cstring>
#include "ephem.h"

#define _def2str(x) #x
#define def2str(x) _def2str(x)

#define HOURMINS 60L
#define DAYMINS (24L * HOURMINS)

using namespace std;

void print_modifed(ostream *out, int minute, int hour, 
		   const char* day, const char* month, const char* dow, 
		   const char* command, const char* tag,
		   const char* mean, const char* var)
{
  out->width(2);
  out->fill('0');
  out->setf(ios::right);
  (*out) << minute << " ";
  
  out->width(2);
  out->fill('0');
  out->setf(ios::right);
  (*out) << hour << " ";

  (*out) << day;
  (*out) << " " << month;
  (*out) << " " << dow;
  (*out) << " " << command;
  (*out) << tag;
  if (NULL!=mean)
    (*out) << " " << mean;
  else if (NULL!=var)
    (*out) << " 0";
  if (NULL!=var)
    (*out) << " " << var;
  (*out) << endl;
}

int wrap(int time)
{
  while (time < 0)        {time += DAYMINS;}
  while (time >= DAYMINS) {time -= DAYMINS;}
  return(time);
}

int noise(int base, int var)
{
  // deterministic
  if (var == 0) return(base);

  // uniform over [-variance.variance]
  var = abs(var);
  // make this calculation with integer math???
  int diff = lrint((2.0 * var * rand()) / RAND_MAX - var);
  return(wrap(base + diff));
}

int timeconv(char *time)
{
  if (time == (char*)NULL)
    return(0);

  if (!strchr(time, ':')) 
    {
      return lrint(atof(time)*HOURMINS);	// rework w/atoi()???
    }
  else
    {
      char* work = new char[strlen(time)+1];
      strcpy(work,time);

      int hour = atoi(strtok(work, ":"));	// get hour
      int min = atoi(strtok(NULL, ":"));	// get minutes
      delete [] work;

      if (hour>=0 && '-' != *time)	 // preserve sign even w/hours==0
	return (hour * HOURMINS + min);
      else
	return (hour * HOURMINS - min);
    }
}

int fixedpoint(int offset,bool intpart,double f)
{
  if (intpart)
    return (int)f;
  return f<0 ?
    -((int)(-0.5 + offset * (f-(int)f))) :
    +((int)( 0.5 + offset * (f-(int)f)));
}

void usage()
{
  cerr << "usage: cronevents";
  cerr << " [crontab_file]";
  cerr << " [lat(deg) long(deg) zone(hrs)]";
  cerr << endl;
  exit(1);
}

int main (int argc, char *argv[])
{
  sleep(1);	// relinquish my timeslice for better randomness later

  ostringstream out;

  istream *in = &cin;
  char *filename = (char*)NULL;

  double latitude  = DLAT;
  double longitude = DLON;
  double timezone;

  if (getenv("LAT") != NULL)
    latitude = atof(getenv("LAT"));

  if (getenv("LON") != NULL)
    longitude = atof(getenv("LON"));

  if (getenv("TOFF") != NULL) {
    char *fakeargv[5]={NULL};
    fakeargv[3]=getenv("TOFF");
    timezone = (double)parseTimezone(4,fakeargv);
  } else
    timezone = (double)parseTimezone(0,NULL);
  
  // valid args:
  //   1==prog
  //   2==prog fspec
  //   4==prog lat lon zone
  //   5==prog fspec lat lon zone
  if ((argc!=1) && (argc!=2) && (argc!=4) && (argc!=5))
    {
      usage();
    }
  
  int arg = 1;

  // process filename
  if ((argc==2) || (argc==5))
    {
      filename = argv[arg];
      ifstream* fin = new ifstream(filename);
      if (!fin->is_open())
	{
	  cerr << "non-existant input file" << endl;
	  usage();
	}
      in = fin;
      arg++;
    }
  
  // process lat, lon, and zone
  if ((arg==1) && (argc==4) ||
      (arg==2) && (argc==5))
    {
      char *endptr = NULL;

      latitude = strtod (argv[arg], &endptr);
      if (endptr == argv[arg]) usage();
      arg++;

      longitude = strtod (argv[arg], &endptr);
      if (endptr == argv[arg]) usage();
      arg++;

      timezone = strtod (argv[arg], &endptr);
      if (endptr == argv[arg]) usage();
      arg++;
    }
  
  time_t wallclock;  time(&wallclock);
  struct tm *tmp = localtime(&wallclock);

  int Y = tmp->tm_year+1900;
  int M = tmp->tm_mon+1;
  int D = tmp->tm_mday; 
  
  // Cheesy RNG initialization, but it works.
  srand(wallclock % 100000);

  int risep;
  float frise;
  float riseaz;
  int setp;
  float fset;
  float setaz;
  int allday;
  
  riseset(latitude, longitude, timezone,
	  Y, M, D,
	  risep, frise, riseaz, setp, fset, setaz, allday);
  
  int rise=lrint(frise*HOURMINS);
  int set=lrint(fset*HOURMINS);

  if (tmp->tm_isdst > 0)		// deal with daylight savings time
    {
      rise += HOURMINS;
      set += HOURMINS;
    }
  
  if (!risep)
    {
      rise = 0;
    }
  
  if (!setp)
    {
      set = DAYMINS;
    }

  unsigned int buflen = 2560;
  char buf[buflen];

  
  while(! in->eof())
    {
      in->getline(buf, buflen, '\n');
      if (buf[0] == '#' || !in->eof() && !strstr(buf,"#x10"))
        {
	  out << buf << endl;
        }
      else if (*buf)
        {
	  char *command;
	  char *token;
	  char *offset;
	  char *variance;
	  char *minute;
	  char *hour;
	  char *day;
	  char *month;
	  char *dow;

	  // split line into two parts to prevent crashing on some comment lines
	  char *cronentry = strtok(buf,"#");
	  char *x10tail = strtok((char *) NULL, "\0");
	  char *side;

	  // process cron part
	  minute = strtok(cronentry, " "); // minute
	  hour = strtok((char *) NULL, " "); // hour

	  day =     strtok((char *) NULL, " "); // day of month
	  month =   strtok((char *) NULL, " "); // month
	  dow =     strtok((char *) NULL, " "); // day of week

	  command = strtok((char *) NULL, "#"); // command

	  // process x10tail
	  token = strtok(x10tail, " "); // pass by the "#x10 " to the real token

	  if (!strcmp(token,"x10#")) {	// x10 conditional that was false the last time
		  // process cron part previously commented out
		  minute = strtok((char *) NULL, " "); // minute
		  hour = strtok((char *) NULL, " "); // hour

		  day =     strtok((char *) NULL, " "); // day of month
		  month =   strtok((char *) NULL, " "); // month
		  dow =     strtok((char *) NULL, " "); // day of week

		  command = strtok((char *) NULL, "#"); // command

		  // process the real x10tail
		  token = strtok((char *) NULL, " "); // pass by the "#x10 " to the real token
	  }

	  token = strtok((char *) NULL, " "); // riseif,setif,rise,set,abs

  	  if (command != (char *) NULL)
	    {
	      int special = 0;	  
	      const char* tag=NULL;
	      char tagspace[buflen];	// room to spare
	      bool conditional = false;
	      
	      if (!strcmp(token, "riseif"))
		{
		  conditional = true;
		  tag = "#x10 riseif ";
		  special = atoi(hour)*HOURMINS + atoi(minute);
		}
	      if (!strcmp(token, "setif"))
		{
		  conditional = true;
		  tag = "#x10 setif ";
		  special = atoi(hour)*HOURMINS + atoi(minute);
		}
	      
	      if (conditional)
	      {
		  side  = strtok((char *) NULL, " "); // before,after
	      }

	    offset = strtok((char *) NULL, " "); // offset
	    variance = strtok((char *) NULL, " "); // randomize
	    // finish strtok()'s above, because timeconv() uses strtok()
	    int mean = timeconv(offset);
	    int var = timeconv(variance);

	      if (conditional && (mean!=0 || var!=0))
	      {
		  special = 0;
	      }

	      if (!strcmp(token, "abs"))
		{
		  tag = "#x10 abs";

		  if ((mean==0) && (var==0))
		    {
		      special = atoi(hour)*HOURMINS + atoi(minute);
		    }
		}
	      if (!strcmp(token, "rise"))
		{
		  tag = "#x10 rise";
		  special = rise;
		}
	      if (!strcmp(token, "set"))
		{
		  tag = "#x10 set";
		  special = set;
		}


	    if (!tag)
	    {
		// space ("# x10") prevents future processing
		strcpy(tagspace,"# x10 unknown: ");
		strcat(tagspace,token);
		tag=tagspace;
	    }


	    // add in the random element
	    int time = noise( wrap(special+mean), var);

	    if (conditional)
	    {
		strcpy(tagspace,tag);
		tag=tagspace;

		if ('b'==*side)	// before
		{
		    strcat(tagspace,"before");
		    if (time > ('r'==*token?rise:set)) // time not before
  			out << " #x10# ";  // preface the line with a comment
		}
		else		// after
		{
		    strcat(tagspace,"after");
		    if (time < ('r'==*token?rise:set)) // time not after
  			out << " #x10# ";  // preface the line with a comment
		}
	    }

	    int hour   = (int) (time / HOURMINS);
	    int minute = (int) (time % HOURMINS);
	    //out << " #x10 special: " << special << "  time: " << time << endl;

	    print_modifed(&out, minute, hour, 
			  day, month, dow,
			  command, tag,
			  offset, variance);
	  }
	}
    }

  // add my special ending tag lines
  //  #x10 oper's: {rise|set|abs|{riseif|setif {before|after}}} [offset [variance]]
  out << " #x10 oper's: {rise|set|abs|{riseif|setif {before|after}}} [offset [variance]]" << endl;
  // order least frequently changed info to most
  //  #x10 @ latN,lonE; {ver} processed 20041114-0305; nodst, day=07:39-17:20
  out << " #x10 @ " <<
    fixedpoint(10,true,latitude) << "." << fixedpoint(10,false,latitude) << "N," <<
    fixedpoint(10,true,longitude) << "." << fixedpoint(10,false,longitude) << "E; " <<
    def2str(VERSION) " processed " << Y;
  out.width(2); out.fill('0'); out << M;
  out.width(2); out.fill('0'); out << D << "-";
  out.width(2); out.fill('0'); out << tmp->tm_hour;
  out.width(2); out.fill('0'); out << tmp->tm_min << "; " <<
    (tmp->tm_isdst > 0 ? "dst" : (tmp->tm_isdst == 0 ? "nodst" : "?dst?")) <<
    ", day=";
  out.width(2); out.fill('0'); out << rise/HOURMINS << ":";
  out.width(2); out.fill('0'); out << rise%HOURMINS << "-";
  out.width(2); out.fill('0'); out << set/HOURMINS << ":";
  out.width(2); out.fill('0'); out << set%HOURMINS << endl;

  if (filename == (char*) NULL)
    {
      cout << out.str();
    }
  else
    {
      ((ifstream*) in)->close();
      sleep(2); // wait a bit so cron doesn't think the file is unchanged
      ofstream fout(filename, ios::out | ios::trunc);
      fout << out.str();
      fout.close();
    }
}
